File: ios/Sources/MultandoSDK/Models/MultandoError.swift
import Foundation

/// Scope of a report-rate-limit breach.
public enum RateLimitScope: String, Sendable {
    case hour
    case day
}

/// Errors thrown by the Multando SDK.
public enum MultandoError: LocalizedError, Sendable {

    /// The API returned a non-2xx status code.
    case apiError(statusCode: Int, message: String)

    /// A network-level failure occurred.
    case networkError(Error)

    /// A request was rejected due to invalid input.
    case validationError(String)

    /// An authentication operation failed (e.g. missing or expired tokens).
    case authError(String)

    /// Response data could not be decoded into the expected type.
    case decodingError(Error)

    /// The caller exceeded the hourly or daily report rate limit.
    /// - Parameters:
    ///   - retryAfter: seconds to wait before retrying, from the `Retry-After`
    ///     header (or the `retry_after_seconds` body field).
    ///   - scope: which window was breached (hour / day).
    case rateLimitExceeded(retryAfter: TimeInterval, scope: RateLimitScope)

    /// The caller tried to report a plate that is still within its cooldown
    /// window (either a per-user duplicate or a plate-volume cap).
    case plateCooldown(plate: String, retryAfterHours: Int)

    public var errorDescription: String? {
        switch self {
        case .apiError(let code, let message):
            return "API error \(code): \(message)"
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        case .validationError(let message):
            return "Validation error: \(message)"
        case .authError(let message):
            return "Auth error: \(message)"
        case .decodingError(let error):
            return "Decoding error: \(error.localizedDescription)"
        case .rateLimitExceeded(_, let scope):
            switch scope {
            case .hour:
                return String(localized: "rate_limit_hour", bundle: .module)
            case .day:
                return String(localized: "rate_limit_day", bundle: .module)
            }
        case .plateCooldown:
            return String(localized: "plate_cooldown", bundle: .module)
        }
    }
}

File: android/multando-sdk/src/main/kotlin/com/multando/sdk/models/MultandoError.kt
package com.multando.sdk.models

/** Scope of a rate-limit window returned by the backend. */
enum class RateLimitScope { HOUR, DAY }

/**
 * Errors thrown by the Multando SDK.
 */
sealed class MultandoError(override val message: String, override val cause: Throwable? = null) : Exception(message, cause) {

    /** The API returned a non-2xx status code. */
    open class ApiError(val statusCode: Int, message: String) :
        MultandoError("API error $statusCode: $message")

    /** A network-level failure occurred. */
    class NetworkError(cause: Throwable) :
        MultandoError("Network error: ${cause.message}", cause)

    /** A request was rejected due to invalid input. */
    class ValidationError(message: String) :
        MultandoError("Validation error: $message")

    /** An authentication operation failed. */
    class AuthError(message: String) :
        MultandoError("Auth error: $message")

    /** Response data could not be decoded into the expected type. */
    class DecodingError(cause: Throwable) :
        MultandoError("Decoding error: ${cause.message}", cause)

    /**
     * The user exceeded an hourly or daily report rate limit.
     *
     * Emitted when the backend returns HTTP 429 with
     * ``error: "rate_limit_exceeded"`` and a ``limit`` of either
     * ``reports_per_hour`` or ``reports_per_day``.
     */
    class RateLimitException(
        message: String,
        val retryAfterSeconds: Long,
        val scope: RateLimitScope,
    ) : ApiError(statusCode = 429, message = message)

    /**
     * The user (or all users) attempted to report a plate within its
     * cooldown window.
     *
     * Emitted when the backend returns HTTP 429 with a plate cooldown
     * ``limit`` (``same_plate_per_user_24h`` / ``plate_reports_24h``).
     */
    class PlateCooldownException(
        message: String,
        val plate: String,
        val retryAfterHours: Int,
    ) : ApiError(statusCode = 429, message = message)
}

/** Convenience aliases mirroring the public task spec naming. */
typealias RateLimitException = MultandoError.RateLimitException
typealias PlateCooldownException = MultandoError.PlateCooldownException
